### PR TITLE
Enable to remove operations by configuration

### PR DESCRIPTION
**Pull request by YalinLi0312**
# Description
Enable to remove operation groups by configurations in readme, and also remove related schemas that not used in other places.

# Checklist

To ensure a quick review and merge, please ensure:
- [ ] The PR has a understandable title and description explaining the _why_ and _what_.
- [ ] The PR is opened in draft if not ready for review yet.
   - If opened in draft, please allocate sufficient time (24 hours) after moving out of draft for review
- [ ] The branch is recent enough to not have merge conflicts upon creation.

# Ready to Land?
- [ ] Build is completely green
   - Submissions with test failures require tracking issue and approval of a CODEOWNER
- [ ] At least one +1 review by a CODEOWNER
- [ ] All -1 reviews are confirmed resolved by the reviewer 
   - Override/Marking reviews stale must be discussed with CODEOWNERS first

### CODE REVIEW

**Comment by pakrym** [comment]
Autorest already has this feature https://github.com/Azure/autorest/blob/6caa31c18ff03173a8f4a945bfe0306c94463aa8/docs/generate/built-in-directives.md#remove-operation

**Comment by YalinLi0312** [comment]
> Autorest already has this feature https://github.com/Azure/autorest/blob/6caa31c18ff03173a8f4a945bfe0306c94463aa8/docs/generate/built-in-directives.md#remove-operation

Does this feature still exist? I simply tried with "remove-operation" directive but get the same result as without it.

File: test/TestProjects/OmitOperationGroups/readme.md [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+``` yaml
+title: ExactMatchInheritance
+require: $(this-folder)/../../../readme.md
+azure-arm: true
+model-namespace: false
+input-file: $(this-folder)/OmitOperationGroups.json
+namespace: OmitOperationGroups
+operation-group-to-resource:
+   Model4s: NonResource
+operation-groups-to-omit: Model1s
```
Comment (allenjzhang) [inline test/TestProjects/OmitOperationGroups/readme.md:14]: nit: put Model1s on new line.

File: test/AutoRest.TestServer.Tests/Mgmt/TestProjects/OmitOperationGroupsTests.cs [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+{
+    public class OmitOperationGroupTests : TestProjectTests
+    {
+        public OmitOperationGroupTests()
+            : base("OmitOperationGroups")
+        {
+        }
+
+        [TestCase(new string[] { "Model1Data", "Model1Update", "Model1ListResult" }, new string[] { "Model2Data", "ModelX", "ModelY", "Model3" })]
+        public void ValidateOperationGroupExistence(string[] notExist, string[] doesExist)
```
Comment (allenjzhang) [inline test/AutoRest.TestServer.Tests/Mgmt/TestProjects/OmitOperationGroupsTests.cs:14]: nit: would have done with (string[] OGs, bool isExist) will be easier to read.

File: test/TestProjects/OmitOperationGroups/readme.md (hunk 1 above)
Comment (allenjzhang) [inline test/TestProjects/OmitOperationGroups/readme.md:14]: nit: on new line.

File: src/AutoRest.CSharp/Mgmt/AutoRest/MgmtOutputLibrary.cs [hunk 3]
```diff
@@ -87,6 +96,186 @@ private void RemoveOperations(CodeModel codeModel)
             }
         }
 
+        private void OmitOperationGroups(CodeModel codeModel, BuildContext<MgmtOutputLibrary> context)
+        {
+            var operationGroupsToOmit = context.Configuration.MgmtConfiguration.OperationGroupsToOmit;
+            if (operationGroupsToOmit != null)
+            {
+                var omitSet = operationGroupsToOmit.ToHashSet();
+                for (int i = 0; i < codeModel.OperationGroups.Count; i++)
```
Comment (allenjzhang) [inline src/AutoRest.CSharp/Mgmt/AutoRest/MgmtOutputLibrary.cs:105]: nit: foreach

File: src/AutoRest.CSharp/Mgmt/AutoRest/MgmtOutputLibrary.cs [hunk 4]
```diff
@@ -44,6 +44,8 @@ private enum ResourceType
         private Dictionary<Operation, MgmtLongRunningOperation>? _longRunningOperations;
         private Dictionary<Operation, NonLongRunningOperation>? _nonLongRunningOperations;
         private Dictionary<string, OperationGroup> _nonResourceOperationGroupMapping;
+        private HashSet<Schema> _schemasToOmit;
```
Comment (allenjzhang) [inline src/AutoRest.CSharp/Mgmt/AutoRest/MgmtOutputLibrary.cs:47]: refCounting would be easier.

**Comment by m-nash** [comment]
> Autorest already has this feature https://github.com/Azure/autorest/blob/6caa31c18ff03173a8f4a945bfe0306c94463aa8/docs/generate/built-in-directives.md#remove-operation

Let me know if we are missing something, but what we want to do is remove an entire operation group, as well as all the lro objects and model objects associated with it without having to specify an enormous list 1 by 1 of everything to omit.

**Comment by pakrym** [comment]
>Let me know if we are missing something, but what we want to do is remove an entire operation group, as well as all the lro objects and model objects associated with it without having to specify an enormous list 1 by 1 of everything to omit.

There is also https://github.com/Azure/autorest.csharp/blob/feature/v3/src/AutoRest.CSharp/Common/Input/SchemaUsageProvider.cs that tracks schema usage and allows you not to generate them.

Maybe we should work with @timotheeguerin to implement a more advanced version of `remove-operation` would be great to avoid having similar features in many places long term.

File: src/AutoRest.CSharp/Mgmt/AutoRest/MgmtOutputLibrary.cs (hunk 3 above)
Comment (YalinLi0312) [reply src/AutoRest.CSharp/Mgmt/AutoRest/MgmtOutputLibrary.cs:105]: Not allowed to use foreach, need to update codeModel.OperationGroups in this loop

File: src/AutoRest.CSharp/Mgmt/AutoRest/MgmtOutputLibrary.cs [hunk 5]
```diff
@@ -137,6 +146,186 @@ private void RemoveOperations(CodeModel codeModel)
             }
         }
 
+        private void OmitOperationGroups(CodeModel codeModel, BuildContext<MgmtOutputLibrary> context)
```
Comment (m-nash) [inline src/AutoRest.CSharp/Mgmt/AutoRest/MgmtOutputLibrary.cs:149]: move all of this logic to a separate class to isolate it and make it easier to refactor later

**Review by m-nash — CHANGES_REQUESTED** [review]

File: src/AutoRest.CSharp/Mgmt/AutoRest/MgmtOutputLibrary.cs [hunk 6]
```diff
@@ -56,18 +58,24 @@ public MgmtOutputLibrary(CodeModel codeModel, BuildContext<MgmtOutputLibrary> co
...
             _childNonResourceOperationGroups = new Dictionary<string, List<OperationGroup>>();
             _nameToTypeProvider = new Dictionary<string, TypeProvider>();
             _nonResourceOperationGroupMapping = new Dictionary<string, OperationGroup>();
+
             _allSchemas = _codeModel.Schemas.Choices.Cast<Schema>()
                 .Concat(_codeModel.Schemas.SealedChoices)
                 .Concat(_codeModel.Schemas.Objects)
                 .Concat(_codeModel.Schemas.Groups);
 
+            _schemasToOmit = new HashSet<Schema>();
```
Comment (allenjzhang) [inline src/AutoRest.CSharp/Mgmt/AutoRest/MgmtOutputLibrary.cs:75]: >_schemasToOmit [](http://example.com/codeflow?start=12&length=14)

nit: remove unused.

File: src/AutoRest.CSharp/Mgmt/Decorator/RemoveOperationGroups.cs [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+        private static HashSet<Schema> _schemasToOmit = new HashSet<Schema>();
+        private static HashSet<Schema> _schemasStillUsed = new HashSet<Schema>();
+
+        public static void OmitOperationGroups(CodeModel codeModel, BuildContext<MgmtOutputLibrary> context)
+        {
+            var operationGroupsToOmit = context.Configuration.MgmtConfiguration.OperationGroupsToOmit;
+            if (operationGroupsToOmit != null)
+            {
+                var omitSet = operationGroupsToOmit.ToHashSet();
+                for (int i = 0; i < codeModel.OperationGroups.Count; i++)
```
Comment (allenjzhang) [inline src/AutoRest.CSharp/Mgmt/Decorator/RemoveOperationGroups.cs:25]: omittedOG = codeModel.OperationGroups.Select(og => omitSet.Contains(og.Key));

codeModel.OperationGroups. = codeModel.OperationGroups.Select(og => !omitSet.Contains(og.Key));

codeModel.OperationGroups.Foreach(og => AddNonOmmited(og));

foreach (var og in ommittedOG)
{
   Line 32-35
}

**Review by m-nash — APPROVED** [review]

**Review by allenjzhang — APPROVED** [review]

**Pull request merged**